File: ImperatorToCK3/Source/Imperator/Provinces/Pop.cpp
#include "Pop.h"
#include "Log.h"
#include "ParserHelpers.h"

Imperator::Pop::Pop(std::istream& theStream, const unsigned long long thePopID) : popID(thePopID)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void Imperator::Pop::registerKeys()
{
	registerKeyword("type", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString typeStr(theStream);
		type = typeStr.getString();
	});
	registerKeyword("culture", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString cultureStr(theStream);
		culture = cultureStr.getString();
	});
	registerKeyword("religion", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString religionStr(theStream);
		religion = religionStr.getString();
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

File: ImperatorToCK3/Imperator/Pops/Pop.cs
using commonItems;
using commonItems.Collections;

namespace ImperatorToCK3.Imperator.Pops;

public class Pop : IIdentifiable<ulong> {
	public ulong Id { get; } = 0;
	public string Type { get; set; } = "";
	public string Culture { get; set; } = "";
	public string Religion { get; set; } = "";
	public Pop(ulong id) {
		Id = id;
	}
	
	public static Pop Parse(string idString, BufferedReader reader) {
		var newPop = new Pop(ulong.Parse(idString));

		var parser = new Parser();
		parser.RegisterKeyword("type", r => newPop.Type = r.GetString());
		parser.RegisterKeyword("culture", r => newPop.Culture = r.GetString());
		parser.RegisterKeyword("religion", r => newPop.Religion = r.GetString());
		parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreAndLogItem);
		parser.ParseStream(reader);
		
		return newPop;
	}
}